contracts/Lens/ComptrollerLens.sol

```diff
@@ -105,113 +126,240 @@ contract ComptrollerLens is ComptrollerLensInterface, ComptrollerErrorReporter,
-                vars.oraclePrice,
+            // borrows += oraclePrice * borrowBalance
+            vars.borrows = mul_ScalarTruncateAddUInt(
+                Exp({ mantissa: vars.oraclePriceMantissa }),
                 vars.borrowBalance,
-                vars.sumBorrowPlusEffects
+                vars.borrows
             );
 
-            // Calculate effects of interacting with vTokenModify
             if (asset == vTokenModify) {
-                // redeem effect
-                // sumBorrowPlusEffects += tokensToDenom * redeemTokens
-                vars.sumBorrowPlusEffects = mul_ScalarTruncateAddUInt(
-                    vars.tokensToDenom,
-                    redeemTokens,
-                    vars.sumBorrowPlusEffects
-                );
+                // redeem effect: weightedVTokenPrice * redeemTokens
+                if (redeemTokens > 0) {
+                    uint256 redeemEffect = mul_ScalarTruncate(weightedVTokenPrice, redeemTokens);
+                    vars.weightedCollateral = vars.weightedCollateral > redeemEffect
+                        ? vars.weightedCollateral - redeemEffect
+                        : 0;
+                }
 
-                // borrow effect
-                // sumBorrowPlusEffects += oraclePrice * borrowAmount
-                vars.sumBorrowPlusEffects = mul_ScalarTruncateAddUInt(
-                    vars.oraclePrice,
+                // borrow effect: oraclePrice * borrowAmount
+                vars.borrows = mul_ScalarTruncateAddUInt(
+                    Exp({ mantissa: vars.oraclePriceMantissa }),
                     borrowAmount,
-                    vars.sumBorrowPlusEffects
+                    vars.borrows
                 );
             }
         }
 
         VAIControllerInterface vaiController = ComptrollerInterface(comptroller).vaiController();
 
         if (address(vaiController) != address(0)) {
-            vars.sumBorrowPlusEffects = add_(vars.sumBorrowPlusEffects, vaiController.getVAIRepayAmount(account));
+            vars.borrows = add_(vars.borrows, vaiController.getVAIRepayAmount(account));
         }
 
-        // These are safe, as the underflow condition is checked first
-        if (vars.sumCollateral > vars.sumBorrowPlusEffects) {
-            return (uint(Error.NO_ERROR), vars.sumCollateral - vars.sumBorrowPlusEffects, 0);
+        (vars.healthFactor, vars.liquidationThresholdAvg, vars.liquidity, vars.shortfall) = _finalizeSnapshot(vars);
+
+        errorCode = NO_ERROR;
+    }
+
+    /**
+     * @notice Finalizes the snapshot of the account's position
+     * @param snapshot The AccountLiquidityLocalVars struct containing the calculated values
+     * @return Returns health factor, average liquidation threshold, liquidity and shortfall.
+     */
+    function _finalizeSnapshot(
+        AccountLiquidityLocalVars memory snapshot
+    ) internal pure returns (uint256, uint256, uint256, uint256) {
+        uint256 healthFactor;
```

Comment:
Would it make sense to initialize this variable to max uint256 (`type(uint256).max`)? Otherwise, with the current implementation, if the user doesn't have any loan, the health factor returned would be zero, that is not correct, IMO

Reply:
[3025a69](https://github.com/VenusProtocol/venus-protocol/pull/604/commits/3025a69bd7b42d0c85c1a4e58144581484c840c3)